### PR TITLE
GitHub actions

### PR DESCRIPTION
**Pull request by dcghub**
### Description of Change ###

This creates a GitHub Actions pipeline that is able to build a nuget package and upload to nuget.org

### Bugs Fixed ###

This fixes bug #1 

### Behavioral Changes ###

This should fix existing build bugs

### CODE REVIEW

File: Xamarin.Essentials/Xamarin.Essentials.csproj [hunk 1]
```diff
@@ -25,12 +25,15 @@
...
     <EnableDefaultCompileItems>false</EnableDefaultCompileItems>
-    <PackageLicenseUrl>https://go.microsoft.com/fwlink/?linkid=868961</PackageLicenseUrl>
+    <PackageLicenseExpression>MIT</PackageLicenseExpression>
     <PackageRequireLicenseAcceptance>true</PackageRequireLicenseAcceptance>
     <PackageProjectUrl>https://go.microsoft.com/fwlink/?linkid=868960</PackageProjectUrl>
     <MDocDocumentationDirectory>$(MSBuildThisFileDirectory)..\docs\en</MDocDocumentationDirectory>
     <Configurations>Debug;Release;Samples;Docs</Configurations>
     <DebugType>portable</DebugType>
+    <IncludeSymbols>true</IncludeSymbols> 
+    <SymbolPackageFormat>snupkg</SymbolPackageFormat>
```
Comment (knocte) [inline Xamarin.Essentials/Xamarin.Essentials.csproj]: why did you include symbols? this might be nice, but was not requested; I would prefer we include this in a different PR

File: Xamarin.Essentials/Xamarin.Essentials.csproj [hunk 2]
```diff
@@ -25,12 +25,15 @@
     <DefineConstants>$(DefineConstants);</DefineConstants>
     <UseFullSemVerForNuGet>false</UseFullSemVerForNuGet>
     <EnableDefaultCompileItems>false</EnableDefaultCompileItems>
-    <PackageLicenseUrl>https://go.microsoft.com/fwlink/?linkid=868961</PackageLicenseUrl>
+    <PackageLicenseExpression>MIT</PackageLicenseExpression>
```
Comment (knocte) [inline Xamarin.Essentials/Xamarin.Essentials.csproj:28]: is this change really necessary for this PR? I think it's just a warning and not an error, right?

File: build.ps1 [hunk 3]
```diff
@@ -24,14 +21,18 @@ Write-Output "Using MSBuild from: $msbuild"
...
-& $msbuild "./Xamarin.Essentials/Xamarin.Essentials.csproj" /t:Pack /p:Configuration=Release /p:ContinuousIntegrationBuild=$cibuild /p:Deterministic=false /p:VersionSuffix=".$env:BUILD_NUMBER-beta"
-if ($lastexitcode -ne 0) { exit $lastexitcode; }
+#& $msbuild "./Xamarin.Essentials/Xamarin.Essentials.csproj" /t:Pack /p:Configuration=Release /p:ContinuousIntegrationBuild=$cibuild /p:Deterministic=false /p:VersionSuffix=".$env:BUILD_NUMBER-beta"
+#if ($lastexitcode -ne 0) { exit $lastexitcode; }
 
 # Copy everything into the output folder
+echo "Copying everything to ./Output"
 Copy-Item "./Xamarin.Essentials/bin/Release" "./Output" -Recurse -Force
 
+get-childitem -recurse .\Output
```
Comment (knocte) [inline build.ps1]: now that the build and upload worked, there's no need anymore for additions like this above (and the echo statements), right? then let's remove them, to minimize the changes between this fork and upstream please

File: Xamarin.Essentials/Xamarin.Essentials.csproj [hunk 4]
```diff
@@ -1,11 +1,11 @@
-﻿<Project Sdk="MSBuild.Sdk.Extras/2.0.31">
+﻿<Project Sdk="MSBuild.Sdk.Extras/2.0.41">
   <PropertyGroup>
     <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT' ">netstandard1.0;netstandard2.0;Xamarin.iOS10;Xamarin.TVOS10;Xamarin.WatchOS10;MonoAndroid60;MonoAndroid70;MonoAndroid71;MonoAndroid80;MonoAndroid81;MonoAndroid90;uap10.0.16299;tizen40;</TargetFrameworks>
     <TargetFrameworks Condition=" '$(OS)' != 'Windows_NT' ">netstandard1.0;netstandard2.0;Xamarin.iOS10;Xamarin.TVOS10;Xamarin.WatchOS10;MonoAndroid60;MonoAndroid70;MonoAndroid71;MonoAndroid80;MonoAndroid81;MonoAndroid90;tizen40;</TargetFrameworks>
     <AssemblyName>Xamarin.Essentials</AssemblyName>
     <RootNamespace>Xamarin.Essentials</RootNamespace>
-    <PackageId>Xamarin.Essentials</PackageId>
-    <PackageIconUrl>https://raw.githubusercontent.com/xamarin/Essentials/master/Assets/xamarin.essentials_128x128.png</PackageIconUrl>
+    <PackageId>TESTXamarin.Essentials</PackageId>
```
Comment (knocte) [inline Xamarin.Essentials/Xamarin.Essentials.csproj]: let's change this to "DotNetEssentials" (I have just set the secret NUGET_API_KEY in the repo to my API key that is able to upload packages to https://www.nuget.org/packages/DotNetEssentials , so let's replace that too in the call to nuget.bat )

File: build.ps1 [hunk 5]
```diff
@@ -24,14 +21,18 @@ Write-Output "Using MSBuild from: $msbuild"
 if ($lastexitcode -ne 0) { exit $lastexitcode; }
 
 # Create the stable NuGet package
-& $msbuild "./Xamarin.Essentials/Xamarin.Essentials.csproj" /t:Pack /p:Configuration=Release /p:ContinuousIntegrationBuild=$cibuild /p:Deterministic=false /p:VersionSuffix=".$env:BUILD_NUMBER"
+echo "Creating package"
+& $msbuild "./Xamarin.Essentials/Xamarin.Essentials.csproj" /t:Pack /p:Configuration=Release /p:ContinuousIntegrationBuild=$cibuild /p:Deterministic=false /p:VersionSuffix=".$date-git$hash"
```
Comment (knocte) [inline build.ps1]: can you tweak the version to include hour:minute too? I want it to be the same format as this sample: 0.10.0-date20191005-0857.git-c03b206 (see https://www.nuget.org/packages/Xamarin.Forms.Background )

File: nuget.bat [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        SET RUNNER=%COMMUNITY%
+    ) ELSE (
+        ECHO fsi.exe not found, is F# installed?
+        EXIT /b 1
+    )
+)
+echo "Running first script"
+%RUNNER% %FSXSCRIPT1% %*
+echo "Running second script"
+%RUNNER% %FSXSCRIPT2% %*
```
Comment (knocte) [inline nuget.bat]: let's rather use this new version of nuget.bat: https://github.com/nblockchain/ZXing.Net.Xamarin/blob/master/nuget.bat (and include fsi.bat)

File: scripts/nuget.fsx [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    let gitHashDefaultShortLength = 7
+    let gitShortHash = gitHash.Substring(0, gitHashDefaultShortLength)
+    let gitSegment = sprintf "git-%s" gitShortHash
+    let finalVersion = sprintf "%s.0-%s.%s"
+                               initialVersion dateSegment gitSegment
+    finalVersion
+
+let PackAndMaybeUpload (packageName: string) =
+    let rootDir = DirectoryInfo(Path.Combine(__SOURCE_DIRECTORY__, ".."))
+    let nugetExe = Path.Combine(rootDir.FullName, "nuget.exe") |> FileInfo
```
Comment (knocte) [inline scripts/nuget.fsx]: you're actually not using the downloaded nuget.exe, are you? (this is a good thing) then we can remove all the lines related to this (including the `Network.DownloadFile`)

File: Xamarin.Essentials/Xamarin.Essentials.csproj (hunk 2 above)
Comment (dcghub) [reply Xamarin.Essentials/Xamarin.Essentials.csproj:28]: I got an error, this is the actual message I got in the build (sorry for the delay by the way):
error NU5125: The 'licenseUrl' element will be deprecated. Consider using the 'license' element instead.

File: build.ps1 (hunk 3 above)
Comment (dcghub) [reply build.ps1]: Done

File: build.ps1 (hunk 5 above)
Comment (dcghub) [reply build.ps1]: Done

File: nuget.bat (hunk 6 above)
Comment (dcghub) [reply nuget.bat]: Fixed.

File: scripts/nuget.fsx (hunk 7 above)
Comment (dcghub) [reply scripts/nuget.fsx]: I didn't even notice that I wasn't using nuget.exe, I will remote it plus all the other unneeded code

File: Xamarin.Essentials/Xamarin.Essentials.csproj (hunk 1 above)
Comment (dcghub) [reply Xamarin.Essentials/Xamarin.Essentials.csproj]: This is a funny one. It is an attempt to circumvent a bug. When the script attempts to upload the package, it does: `dotnet nuget push Output\*.nupkg`. It turns out that when you specify packages with a `*`wildcard and finds a package, it automatically searches for a .snupkg - and exits with failure if it doesn't find it. This is a known bug that for some reason hasn't been fixed, and it does not happen when you specify the name of the package directly.

The only solution possible (other than generate the .snupkg) is to specify the package name directly, by searching the file and extracting it in scripts/nuget.fsx. I chose the simplest solution, (because I don't know F# and because I expect the nuget bug will get solved some day), but I can do the more complex one if necessary.

File: Xamarin.Essentials/Xamarin.Essentials.csproj (hunk 4 above)
Comment (dcghub) [reply Xamarin.Essentials/Xamarin.Essentials.csproj]: Done.

File: Xamarin.Essentials/Xamarin.Essentials.csproj (hunk 2 above)
Comment (knocte) [reply Xamarin.Essentials/Xamarin.Essentials.csproj:28]: was this error really preventing the build to succeed?

File: Xamarin.Essentials/Xamarin.Essentials.csproj (hunk 1 above)
Comment (knocte) [reply Xamarin.Essentials/Xamarin.Essentials.csproj]: well, it was you who introduced the wildcard :) I mean, the original nuget.fsx script you copied this from didn't have a wildcard, precisely for this reason! I'd prefer you remove the wildcard and remove the snupkg generation too.

File: Xamarin.Essentials/Xamarin.Essentials.csproj (hunk 4 above)
Comment (knocte) [reply Xamarin.Essentials/Xamarin.Essentials.csproj]: you didn't change this to DotNetEssentials, you left it as Xamarin.Essentials

Comment (dcghub) [reply Xamarin.Essentials/Xamarin.Essentials.csproj]: Sorry, I left the original name insted of the new one you suggested.

File: Xamarin.Essentials/Xamarin.Essentials.csproj [hunk 8]
```diff
@@ -8,7 +8,7 @@
     <PackageIcon>Assets\xamarin.essentials_128x128.png</PackageIcon>
     <Summary>Xamarin.Essentials: a kit of essential API's for your apps</Summary>
     <PackageTags>xamarin, windows, ios, android, xamarin.forms, watchos, tvos, toolkit, Xamarin.Essentials, kit</PackageTags>
-    <Title>Xamarin.Essentials</Title>
+    <Title>DotNetEssentials</Title>
```
Comment (knocte) [inline Xamarin.Essentials/Xamarin.Essentials.csproj]: @dcghub you changed Title, but I think you have to change PackageId

Comment (dcghub) [reply Xamarin.Essentials/Xamarin.Essentials.csproj]: Apparently today it's not being my brightest day...
Should RootNamespace and/or AssemblyName also be changed?

Comment (knocte) [reply Xamarin.Essentials/Xamarin.Essentials.csproj]: nope, we just need to change PackageId for the nuget upload to the proper package

File: Xamarin.Essentials/Xamarin.Essentials.csproj (hunk 1 above)
Comment (dcghub) [reply Xamarin.Essentials/Xamarin.Essentials.csproj]: This one is going to take me more time than expected, as the old script code can't be reused and I have to learn the basics of F# to attempt to pass the string filename to nupkg; unfortunately F# is not very kind to newcomers...

File: Xamarin.Essentials/Xamarin.Essentials.csproj (hunk 8 above)
Comment (knocte) [reply Xamarin.Essentials/Xamarin.Essentials.csproj]: please leavee the title as it was

File: Xamarin.Essentials/Xamarin.Essentials.csproj (hunk 1 above)
Comment (knocte) [reply Xamarin.Essentials/Xamarin.Essentials.csproj]: try this:
```
        let outputDir = Path.Combine(Directory.GetCurrentDirectory(), "Output") |> DirectoryInfo
        let nupkgFile = outputDir.EnumerateFiles().Single(fun file -> file.FullName.EndsWith ".nupkg")
```

and later:

```
        let Arguments = sprintf "nuget push %s -k %s -s %s"
                            nupkgFile.FullName nugetApiKey defaultNugetFeedUrl
```

Comment (knocte) [reply Xamarin.Essentials/Xamarin.Essentials.csproj]: (not forgetting to put this at the top: 
```
open System.IO
open System.Linq
```
)

File: Xamarin.Essentials/Xamarin.Essentials.csproj (hunk 2 above)
Comment (dcghub) [reply Xamarin.Essentials/Xamarin.Essentials.csproj:28]: It appears so:
2019-10-17T23:12:56.9871907Z   C:\Program Files\dotnet\sdk\3.0.100\Sdks\NuGet.Build.Tasks.Pack\buildCrossTargeting\NuGet.Build.Tasks.Pack.targets(198,5): error : Enabling license acceptance requires a license or a licenseUrl to be specified. The licenseUrl will be deprecated, consider using the license metadata. [D:\a\DotNetEssentials\DotNetEssentials\Xamarin.Essentials\Xamarin.Essentials.csproj]
2019-10-17T23:12:56.9872757Z     16 Warning(s)
2019-10-17T23:12:56.9873018Z     1 Error(s)
2019-10-17T23:12:56.9873376Z Time Elapsed 00:00:17.11
2019-10-17T23:12:57.1364769Z ##[error]Process completed with exit code 1.

File: scripts/nuget.fsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+                Arguments = sprintf "nuget push %s -k %s -s %s"
+                                    nupkgFile.FullName nugetApiKey defaultNugetFeedUrl
+            }
+        Console.WriteLine "Pushing package..."
+        Process.SafeExecute (nugetPushCmd, Echo.All) |> ignore
+        Console.WriteLine "Pushing completed!"
+
+
+Console.WriteLine "Packaging..."
+PackAndMaybeUpload "Xamarin.Essentials"
```
Comment (knocte) [inline scripts/nuget.fsx]: nitpick: this function doesn't do pack anymore, so please rename it to MaybeUpload

File: build.ps1 [hunk 10]
```diff
@@ -1,19 +1,16 @@
 $ErrorActionPreference = "Stop"
 
 $cibuild = "false"
-
-# Make sure that we have something on non-bots
-if (!$env:BUILD_NUMBER) {
-    $env:BUILD_NUMBER = "0"
-}
```
Comment (knocte) [inline build.ps1:8]: you don't actually need to remove the lines above, right? the less changes we have compared to upstream, the better

File: scripts/nuget.fsx (hunk 9 above)
Comment (dcghub) [reply scripts/nuget.fsx]: Done!

File: build.ps1 (hunk 10 above)
Comment (dcghub) [reply build.ps1:8]: I will restore it, but note that this variable is no longer used in the script.

File: Xamarin.Essentials/Xamarin.Essentials.csproj (hunk 1 above)
Comment (dcghub) [reply Xamarin.Essentials/Xamarin.Essentials.csproj]: Done, many thanks for your help.

File: Xamarin.Essentials/Xamarin.Essentials.csproj (hunk 8 above)
Comment (dcghub) [reply Xamarin.Essentials/Xamarin.Essentials.csproj]: This is done too

**Comment by knocte** [comment]
Superseded by https://github.com/nblockchain/DotNetEssentials/pull/3
